File: crates/hyper_rhi/src/vulkan/render_pipeline.rs
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
*/

use std::ffi::CStr;

use ash::vk;

use crate::{
    render_pipeline::RenderPipelineDescriptor,
    shader_compiler::{self, OutputApi, ShaderStage},
    vulkan::GraphicsDevice,
};

pub struct RenderPipeline {
    pipeline: vk::Pipeline,
    graphics_device: GraphicsDevice,
}

impl RenderPipeline {
    pub(crate) fn new(
        graphics_device: &GraphicsDevice,
        descriptor: &RenderPipelineDescriptor,
    ) -> Self {
        // NOTE: Maybe instead of failure compile shaders on its own for multiple pipelines and don't panic
        let vertex_shader = shader_compiler::compile(
            descriptor.vertex_shader,
            ShaderStage::Vertex,
            OutputApi::Vulkan,
        )
        .expect("failed to compile vertex shader");
        let vertex_shader_module =
            Self::create_shader_module(graphics_device, &vertex_shader, "vertex");

        let pixel_shader = shader_compiler::compile(
            descriptor.pixel_shader,
            ShaderStage::Pixel,
            OutputApi::Vulkan,
        )
        .expect("failed to compile pixel shader");
        let pixel_shader_module =
            Self::create_shader_module(graphics_device, &pixel_shader, "pixel");

        let pipeline =
            Self::create_pipeline(graphics_device, vertex_shader_module, pixel_shader_module);

        unsafe {
            graphics_device
                .device()
                .destroy_shader_module(vertex_shader_module, None);
            graphics_device
                .device()
                .destroy_shader_module(pixel_shader_module, None);
        }

        Self {
            pipeline,
            graphics_device: graphics_device.clone(),
        }
    }

    fn create_shader_module(
        graphics_device: &GraphicsDevice,
        code: &[u8],
        stage: &str,
    ) -> vk::ShaderModule {
        let (prefix, code, suffix) = unsafe { code.align_to::<u32>() };
        if !prefix.is_empty() || !suffix.is_empty() {
            panic!("failed to align {} shader", stage);
        }

        let create_info = vk::ShaderModuleCreateInfo::default().code(code);

        let shader_module = unsafe {
            graphics_device
                .device()
                .create_shader_module(&create_info, None)
        }
        .expect(&format!("failed to create {} shader module", stage));
        shader_module
    }

    fn create_pipeline(
        graphics_device: &GraphicsDevice,
        vertex_shader: vk::ShaderModule,
        pixel_shader: vk::ShaderModule,
    ) -> vk::Pipeline {
        let entry_name = unsafe { CStr::from_bytes_with_nul_unchecked(b"main\0") };

        let vertex_shader_stage_create_info = vk::PipelineShaderStageCreateInfo::default()
            .stage(vk::ShaderStageFlags::VERTEX)
            .module(vertex_shader)
            .name(entry_name);

        let pixel_shader_stage_create_info = vk::PipelineShaderStageCreateInfo::default()
            .stage(vk::ShaderStageFlags::FRAGMENT)
            .module(pixel_shader)
            .name(entry_name);

        let shader_stages = [
            vertex_shader_stage_create_info,
            pixel_shader_stage_create_info,
        ];
        let dynamic_states = [vk::DynamicState::VIEWPORT, vk::DynamicState::SCISSOR];

        let dynamic_state_create_info =
            vk::PipelineDynamicStateCreateInfo::default().dynamic_states(&dynamic_states);

        let vertex_input_state_create_info = vk::PipelineVertexInputStateCreateInfo::default()
            .vertex_binding_descriptions(&[])
            .vertex_attribute_descriptions(&[]);

        let input_assembly_state_create_info = vk::PipelineInputAssemblyStateCreateInfo::default()
            .topology(vk::PrimitiveTopology::TRIANGLE_LIST)
            .primitive_restart_enable(false);

        let tessellation_state_create_info =
            vk::PipelineTessellationStateCreateInfo::default().patch_control_points(0);

        let viewport_state_create_info = vk::PipelineViewportStateCreateInfo::default()
            .viewport_count(1)
            .scissor_count(1);

        let rasterization_state_create_info = vk::PipelineRasterizationStateCreateInfo::default()
            .depth_clamp_enable(false)
            .rasterizer_discard_enable(false)
            .polygon_mode(vk::PolygonMode::FILL)
            .line_width(1.0)
            .cull_mode(vk::CullModeFlags::NONE)
            .front_face(vk::FrontFace::COUNTER_CLOCKWISE)
            .depth_bias_enable(false)
            .depth_bias_constant_factor(0.0)
            .depth_bias_clamp(0.0)
            .depth_bias_slope_factor(0.0);

        // TODO: Add multisampling
        let multisample_state_create_info = vk::PipelineMultisampleStateCreateInfo::default()
            .sample_shading_enable(false)
            .rasterization_samples(vk::SampleCountFlags::TYPE_1)
            .min_sample_shading(1.0)
            .sample_mask(&[])
            .alpha_to_coverage_enable(false)
            .alpha_to_one_enable(false);

        // TODO: Add stencil
        let depth_stencil_state_create_info = vk::PipelineDepthStencilStateCreateInfo::default()
            .depth_test_enable(false)
            .depth_write_enable(false)
            .depth_compare_op(vk::CompareOp::NEVER)
            .depth_bounds_test_enable(false)
            .stencil_test_enable(false)
            .front(vk::StencilOpState::default())
            .back(vk::StencilOpState::default())
            .min_depth_bounds(0.0)
            .max_depth_bounds(0.0);

        let color_blend_attachment_state_create_info =
            vk::PipelineColorBlendAttachmentState::default()
                .blend_enable(false)
                .src_color_blend_factor(vk::BlendFactor::ONE)
                .dst_color_blend_factor(vk::BlendFactor::ZERO)
                .color_blend_op(vk::BlendOp::ADD)
                .src_alpha_blend_factor(vk::BlendFactor::ONE)
                .dst_alpha_blend_factor(vk::BlendFactor::ZERO)
                .alpha_blend_op(vk::BlendOp::ADD)
                .color_write_mask(vk::ColorComponentFlags::RGBA);

        let attachments = &[color_blend_attachment_state_create_info];

        let color_blend_state_create_info = vk::PipelineColorBlendStateCreateInfo::default()
            .logic_op_enable(false)
            .logic_op(vk::LogicOp::NO_OP)
            .attachments(attachments)
            .blend_constants([0.0; 4]);

        let color_attachment_formats = &[vk::Format::B8G8R8A8_UNORM];
        let mut rendering_create_info = vk::PipelineRenderingCreateInfoKHR::default()
            .color_attachment_formats(color_attachment_formats);

        let create_info = vk::GraphicsPipelineCreateInfo::default()
            .push_next(&mut rendering_create_info)
            .stages(&shader_stages)
            .vertex_input_state(&vertex_input_state_create_info)
            .input_assembly_state(&input_assembly_state_create_info)
            .tessellation_state(&tessellation_state_create_info)
            .viewport_state(&viewport_state_create_info)
            .rasterization_state(&rasterization_state_create_info)
            .multisample_state(&multisample_state_create_info)
            .depth_stencil_state(&depth_stencil_state_create_info)
            .color_blend_state(&color_blend_state_create_info)
            .dynamic_state(&dynamic_state_create_info)
            .layout(graphics_device.pipeline_layout())
            .render_pass(vk::RenderPass::null())
            .subpass(0)
            .base_pipeline_handle(vk::Pipeline::null())
            .base_pipeline_index(-1);

        let pipeline = unsafe {
            graphics_device
                .device()
                .create_graphics_pipelines(vk::PipelineCache::null(), &[create_info], None)
                .expect("failed to create render pipeline")[0]
        };
        pipeline
    }

    pub(crate) fn pipeline(&self) -> vk::Pipeline {
        self.pipeline
    }
}

impl Drop for RenderPipeline {
    fn drop(&mut self) {
        unsafe {
            self.graphics_device
                .device()
                .destroy_pipeline(self.pipeline, None);
        }
    }
}

File: hyper_rhi/src/hyper_rhi/vulkan/render_pipeline.cpp
/*
 * Copyright (c) 2025-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/vulkan/render_pipeline.hpp"

#include <array>

#include "hyper_rhi/validation.hpp"
#include "hyper_rhi/vulkan/graphics_device.hpp"
#include "hyper_rhi/vulkan/pipeline_layout.hpp"
#include "hyper_rhi/vulkan/render_conversion.hpp"
#include "hyper_rhi/vulkan/shader.hpp"
#include "hyper_rhi/vulkan/utils.hpp"

namespace he
{
    VulkanRenderPipeline::VulkanRenderPipeline(
        VulkanGraphicsDevice &graphics_device, const RenderPipelineDescriptor &desc)
        : RenderPipeline(desc)
        , m_graphics_device(graphics_device)
    {
        validate_render_pipeline_descriptor(desc);

        const RefPtr<VulkanShader> vertex_shader = cast_ref<VulkanShader>(desc.vertex_shader);
        const VkPipelineShaderStageCreateInfo vertex_pipeline_shader_stage_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .stage = VK_SHADER_STAGE_VERTEX_BIT,
            .module = vertex_shader->raw(),
            .pName = vertex_shader->entry().data(),
            .pSpecializationInfo = nullptr,
        };

        const RefPtr<VulkanShader> fragment_shader = cast_ref<VulkanShader>(desc.fragment_shader);
        const VkPipelineShaderStageCreateInfo fragment_pipeline_shader_stage_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .stage = VK_SHADER_STAGE_FRAGMENT_BIT,
            .module = fragment_shader->raw(),
            .pName = fragment_shader->entry().data(),
            .pSpecializationInfo = nullptr,
        };

        const std::array<VkPipelineShaderStageCreateInfo, 2> shader_stage_create_infos = {
            vertex_pipeline_shader_stage_create_info,
            fragment_pipeline_shader_stage_create_info,
        };

        constexpr VkPipelineVertexInputStateCreateInfo vertex_input_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .vertexBindingDescriptionCount = 0,
            .pVertexBindingDescriptions = nullptr,
            .vertexAttributeDescriptionCount = 0,
            .pVertexAttributeDescriptions = nullptr,
        };

        const VkPipelineInputAssemblyStateCreateInfo input_assembly_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_INPUT_ASSEMBLY_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .topology = map_primitive_topology(desc.primitive_state.topology),
            .primitiveRestartEnable = false,
        };

        constexpr VkPipelineTessellationStateCreateInfo tessellation_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_TESSELLATION_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .patchControlPoints = 0,
        };

        constexpr VkPipelineViewportStateCreateInfo viewport_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_VIEWPORT_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .viewportCount = 1,
            .pViewports = nullptr,
            .scissorCount = 1,
            .pScissors = nullptr,
        };

        VkPipelineRasterizationStateCreateInfo rasterization_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_RASTERIZATION_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .depthClampEnable = false,
            .rasterizerDiscardEnable = false,
            .polygonMode = map_polygon_mode(desc.primitive_state.polygon_mode),
            .cullMode = map_cull_mode(desc.primitive_state.cull_mode),
            .frontFace = map_front_face(desc.primitive_state.front_face),
            .depthBiasEnable = false,
            .depthBiasConstantFactor = 0.0f,
            .depthBiasClamp = 0.0f,
            .depthBiasSlopeFactor = 0.0f,
            .lineWidth = 1.0,
        };

        if (desc.depth_stencil_state.has_value())
        {
            const DepthStencilState &depth_stencil_state = desc.depth_stencil_state.value();
            rasterization_state_create_info.depthBiasEnable = depth_stencil_state.depth_bias_state.enable;
            rasterization_state_create_info.depthBiasConstantFactor = depth_stencil_state.depth_bias_state.constant;
            rasterization_state_create_info.depthBiasClamp = depth_stencil_state.depth_bias_state.clamp;
            rasterization_state_create_info.depthBiasSlopeFactor = depth_stencil_state.depth_bias_state.slope;
        }

        const VkPipelineMultisampleStateCreateInfo multisample_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .rasterizationSamples = map_sample_count(desc.multisample_state.sample_count),
            .sampleShadingEnable = true,
            .minSampleShading = 0.2f,
            .pSampleMask = nullptr,
            .alphaToCoverageEnable = false,
            .alphaToOneEnable = false,
        };

        // FIXME: Add depth bounds & stencil
        VkPipelineDepthStencilStateCreateInfo depth_stencil_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .depthTestEnable = false,
            .depthWriteEnable = false,
            .depthCompareOp = VK_COMPARE_OP_NEVER,
            .depthBoundsTestEnable = false,
            .stencilTestEnable = false,
            .front = {},
            .back = {},
            .minDepthBounds = 0.0,
            .maxDepthBounds = 1.0,
        };

        if (desc.depth_stencil_state.has_value())
        {
            const DepthStencilState &depth_stencil_state = desc.depth_stencil_state.value();
            depth_stencil_state_create_info.depthTestEnable = depth_stencil_state.depth_test_enable;
            depth_stencil_state_create_info.depthWriteEnable = depth_stencil_state.depth_write_enable;

            const VkCompareOp depth_compare_operation
                = map_compare_operation(depth_stencil_state.depth_compare_operation);
            depth_stencil_state_create_info.depthCompareOp = depth_compare_operation;
        }

        std::vector<VkPipelineColorBlendAttachmentState> color_blend_attachment_states;
        for (const ColorAttachmentState &color_attachment_state : desc.color_attachment_states)
        {
            const BlendState &blend_state = color_attachment_state.blend_state;

            const VkPipelineColorBlendAttachmentState color_blend_attachment_state = {
                .blendEnable = blend_state.enable,
                .srcColorBlendFactor = map_blend_factor(blend_state.src_factor),
                .dstColorBlendFactor = map_blend_factor(blend_state.dst_factor),
                .colorBlendOp = map_blend_operation(blend_state.operation),
                .srcAlphaBlendFactor = map_blend_factor(blend_state.alpha_src_factor),
                .dstAlphaBlendFactor = map_blend_factor(blend_state.alpha_dst_factor),
                .alphaBlendOp = map_blend_operation(blend_state.alpha_operation),
                .colorWriteMask = map_color_writes(blend_state.color_writes),
            };

            color_blend_attachment_states.push_back(color_blend_attachment_state);
        }

        const VkPipelineColorBlendStateCreateInfo color_blend_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_COLOR_BLEND_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .logicOpEnable = false,
            .logicOp = VK_LOGIC_OP_NO_OP,
            .attachmentCount = static_cast<u32>(color_blend_attachment_states.size()),
            .pAttachments = color_blend_attachment_states.data(),
            .blendConstants = { 0.0, 0.0, 0.0, 0.0 },
        };

        constexpr std::array<VkDynamicState, 2> dynamic_states = {
            VK_DYNAMIC_STATE_VIEWPORT,
            VK_DYNAMIC_STATE_SCISSOR,
        };

        const VkPipelineDynamicStateCreateInfo dynamic_state_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_DYNAMIC_STATE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .dynamicStateCount = static_cast<u32>(dynamic_states.size()),
            .pDynamicStates = dynamic_states.data(),
        };

        std::vector<VkFormat> color_attachment_formats;
        for (const ColorAttachmentState &color_attachment_state : desc.color_attachment_states)
        {
            color_attachment_formats.push_back(map_format(color_attachment_state.format));
        }

        const VkFormat depth_attachment_format = desc.depth_stencil_state.has_value()
            ? map_format(desc.depth_stencil_state.value().depth_format)
            : VK_FORMAT_UNDEFINED;
        const VkPipelineRenderingCreateInfo rendering_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_RENDERING_CREATE_INFO,
            .pNext = nullptr,
            .viewMask = 0,
            .colorAttachmentCount = static_cast<u32>(color_attachment_formats.size()),
            .pColorAttachmentFormats = color_attachment_formats.data(),
            .depthAttachmentFormat = depth_attachment_format,
            .stencilAttachmentFormat = VK_FORMAT_UNDEFINED,
        };

        const VkGraphicsPipelineCreateInfo graphics_pipeline_create_info = {
            .sType = VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO,
            .pNext = &rendering_create_info,
            .flags = 0,
            .stageCount = static_cast<u32>(shader_stage_create_infos.size()),
            .pStages = shader_stage_create_infos.data(),
            .pVertexInputState = &vertex_input_state_create_info,
            .pInputAssemblyState = &input_assembly_state_create_info,
            .pTessellationState = &tessellation_state_create_info,
            .pViewportState = &viewport_state_create_info,
            .pRasterizationState = &rasterization_state_create_info,
            .pMultisampleState = &multisample_state_create_info,
            .pDepthStencilState = &depth_stencil_state_create_info,
            .pColorBlendState = &color_blend_state_create_info,
            .pDynamicState = &dynamic_state_create_info,
            .layout = cast_ref<VulkanPipelineLayout>(desc.layout)->raw(),
            .renderPass = VK_NULL_HANDLE,
            .subpass = 0,
            .basePipelineHandle = VK_NULL_HANDLE,
            .basePipelineIndex = -1,
        };

        HE_VK_CHECK(
            vkCreateGraphicsPipelines(
                m_graphics_device.device(), VK_NULL_HANDLE, 1, &graphics_pipeline_create_info, nullptr, &m_raw),
            "Failed to create vulkan graphics pipeline");
        HE_ASSERT(m_raw != VK_NULL_HANDLE);
    }

    VulkanRenderPipeline::~VulkanRenderPipeline() { vkDestroyPipeline(m_graphics_device.device(), m_raw, nullptr); }
} // namespace he
